### PR TITLE
Ability to derive Startup from base class 

### PR DESCRIPTION
**Pull request by dtkujawski**
Ability to derive Startup from base class and have "Configure" and "ConfigureServices" be invoked from the base class when present.

### CODE REVIEW

**Comment by msotclas** [comment]
Hi __@dtkujawski__, I'm your friendly neighborhood Microsoft Open Technologies, Inc. Pull Request Bot (You can call me MSOTBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft Open Technologies, Inc. and real humans are currently evaluating your PR.</p>

TTYL, MSOTBOT;

**Comment by dtkujawski** [comment]
Created pull request from issue:
https://github.com/aspnet/Hosting/issues/105

**Comment by glennc** [comment]
@davidfowl @GrabYourPitchforks can one of you take a look at this?

File: src/Microsoft.AspNet.Hosting/Startup/StartupLoader.cs [hunk 1]
```diff
@@ -29,18 +29,18 @@ private MethodInfo FindMethod(Type startupType, string methodName, string enviro
         {
             var methodNameWithEnv = string.Format(CultureInfo.InvariantCulture, methodName, environmentName);
             var methodNameWithNoEnv = string.Format(CultureInfo.InvariantCulture, methodName, "");
-            var methodInfo = startupType.GetTypeInfo().GetDeclaredMethod(methodNameWithEnv)
-                ?? startupType.GetTypeInfo().GetDeclaredMethod(methodNameWithNoEnv);
-            if (methodInfo == null)
+			var methodInfo = startupType.GetMethod(methodNameWithEnv, BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static)
```
Comment (Eilon) [inline src/Microsoft.AspNet.Hosting/Startup/StartupLoader.cs]: OMG tabs! Please change to spaces :smile:

**Comment by Eilon** [comment]
Overall this does look good to me. I just have to wonder, though, could there have been any legitimate reason why this wasn't allowed before? Confusion among multiple startup classes?

**Comment by glennc** [comment]
That was my thought as well. I asked @davidfowl about it briefly and he didn't seem to think so.

**Comment by davidfowl** [comment]
Yep, I can't think of a reason block this scenario. You could have a bunch of Startup base classes that setup default services for your company or something.

**Comment by dtkujawski** [comment]
Sorry for the tabs, BTW.  I wasn't able to use Visual Studio due to being on CTP 5 and some issues compiling this code and my trusty text editor let me down!

Is there some way I can change the pull request and get that corrected?  I'm quite new with github, but I'm willing to do whatever I can to satisfy your coding guidelines.

**Comment by davidfowl** [comment]
@dtkujawski just update the code and push to the branch. The pull request is just a view of your branch.

**Comment by dtkujawski** [comment]
Thanks @davidfowl, I've pushed an update which cleans that tab mess up.

File: test/Microsoft.AspNet.Hosting.Tests/Fakes/StartupBase.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNet.Builder;
+using Microsoft.Framework.DependencyInjection;
+using Microsoft.Framework.DependencyInjection.Fallback;
```
Comment (Eilon) [inline test/Microsoft.AspNet.Hosting.Tests/Fakes/StartupBase.cs]: Don't need this namespace. Might be a few other extraneous namespaces as well (Builder?).

File: test/Microsoft.AspNet.Hosting.Tests/Fakes/StartupBase.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+using Microsoft.Framework.OptionsModel;
+using System;
+
+namespace Microsoft.AspNet.Hosting.Fakes
+{
+    public class StartupBase
+    {
+        public StartupBase()
```
Comment (Eilon) [inline test/Microsoft.AspNet.Hosting.Tests/Fakes/StartupBase.cs]: Can remove this ctor?

File: test/Microsoft.AspNet.Hosting.Tests/Fakes/StartupBase.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        public void ConfigureBaseClassServices(IServiceCollection services)
+        {
+            services.AddOptions();
+            services.Configure<FakeOptions>(o =>
+            {
+                o.Configured = true;
+                o.Environment = "BaseClass";
+            });
+        }
+
```
Comment (Eilon) [inline test/Microsoft.AspNet.Hosting.Tests/Fakes/StartupBase.cs]: And empty space while we're at it :smile:

**Comment by Eilon** [comment]
@dtkujawski this is looking really good - just a few small cleanup comments.

**Comment by dtkujawski** [comment]
Thanks - I made those changes.  On a extremely positive note: I re-fetched the entire code base and am no able to use Visual Studio again to write code instead of notepad!  :)   That makes things so much easier!

**Comment by Eilon** [comment]
Great to hear!

@davidfowl looks good to you?

**Comment by davidfowl** [comment]
Yep!

**Comment by Eilon** [comment]
@davidfowl wanna merge?

**Comment by Eilon** [comment]
CI passed, thanks for the contribution!

**Pull request merged**